ARM64-SVE: Delay free all ops within conditional select

a74nh opened the pull request:
Fixes #106864
Fixes #106867

Embedded ops may be prefixed by a MOVPFRX instruction. If so, then the embedded op can't use the reuse a source register as the destination. Ensure all the inputs are marked as delay free.

This solves the case where an embedded mask operation is passed two identical inputs.

a74nh committed:
ARM64-SVE: Delay free all ops within conditional select

a74nh commented:
@dotnet/arm64-contrib @kunalspathak

amanasifkhalid reviewed (approved):
LGTM, thanks!

tannergooding commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -2040,26 +2040,16 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
-            GenTree* prefUseNode = embOp2Node->Op(prefUseOpNum);
+            // The embedded op may be prefixed by a MOVPFRX instruction. If so, then the embedded
+            // op can't use the reuse a source register as the destination. Ensure all the inputs
+            // are marked as delay free.
```
We shouldn't be delay freeing everything right, because aren't we are expecting that `op1` and the `destReg` be the same to avoid additional unnecessary moves?

That is, take `add (Vectors, predicated)` for example. In that case we have:
```
ADD <Zdn>.<T>, <Pg>/M, <Zdn>.<T>, <Zm>.<T>
```

So in the scenario that we don't need `movprfx`, then `tgtPrefUse == op1` (therefore `op1` is not delay free) and `op2` does not need to be delay free. However, in the case we do need `movprfx`, then we still want `tgtPrefUse == op1` (and therefore `op1` is not delay free) but we do require that `op2` be delay free, due to the `movprfx` requirements. -- If we're marking `op1` as delay free, then it won't be assign the same register as the destination, and that means we need to insert an additional `mov tgt, op1` which should be preventing us from emitting `movprfx` anyways (and we should be emitting `add (unpredicated); sel` instead).

For .NET 9, we likely want to pessimize and assume that `movprfx` will also be needed, just because that's the simpler fix and we're already in RC2. However, we should ensure that an issue exists for .NET 10 and we fix the codegen to do the optimal thing here before we ship the feature as stable.

That includes handling (in .NET 10) cases like `add` which are commutative, where we want to determine which of `op1` or `op2` is more likely to be selected as the destination (normally done via `getTgtPrefOperands` or similar) and swap them if `op2` is more likely. -- That is, I imagine we really want something like xarch's `LinearScan::BuildRMWUses` for the SVE RMW instructions here.

a74nh replied:
> For .NET 9, we likely want to pessimize and assume that `movprfx` will also be needed, just because that's the simpler fix and we're already in RC2. However, we should ensure that an issue exists for .NET 10 and we fix the codegen to do the optimal thing here before we ship the feature as stable.

Agreed and will do.


> We shouldn't be delay freeing everything right, because aren't we are expecting that `op1` and the `destReg` be the same to avoid additional unnecessary moves?

> However, in the case we do need `movprfx`, then we still want `tgtPrefUse == op1` (and therefore `op1` is not delay free) but we do require that `op2` be delay free, due to the `movprfx` requirements.

Which means the original code was doing the right thing. `Op1` used `BuildUse()` and `Op2` used `BuildDelayFreeUses()`.

But, the jit was still allowing `Op1` and `Op2` to have the same registers.

With the code in HEAD we get:

```
Generating: N029 (  4,  3) [000026] ---XG+-----                   t26 = *  IND       simd16 REG d16 <l:$2c1, c:$2c2>
IN000a:             ldr     q16, [x0, #0x08]
                                                                        /--*  t26    simd16 
Generating: N031 (  4,  3) [000052] DA-XG------                         *  STORE_LCL_VAR simd16 V07 cse0         d:1 d16 REG d16 $VN.Void
							V07 in reg d16 is becoming live  [000052]
							Live regs: 0000000000000001 {x0} + {d16} => 0001000000000001 {x0 d16}
							Live vars after [000052]: {V03} +{V07} => {V03 V07}
Generating: N033 (  1,  1) [000053] -----------                   t53 =    LCL_VAR   simd16 V07 cse0         u:1 d16 REG d16 <l:$2c1, c:$2c2>
Generating: N035 (  1,  1) [000055] -----------                   t55 =    LCL_VAR   simd16 V07 cse0         u:1 d16 (last use) REG d16 <l:$2c0, c:$300>
Generating: N037 (???,???) [000073] -----------                   t73 =    HWINTRINSIC mask   short CreateTrueMaskAll REG p0
IN000b:             ptrue   p0.h
Generating: N039 (???,???) [000074] -c---------                   t74 =    CNS_VEC   simd16<0x00000000, 0x00000000, 0x00000000, 0x00000000> REG NA
                                                                        /--*  t53    simd16 
                                                                        +--*  t55    simd16 
Generating: N041 (  7,  6) [000034] -A-XG+-----                   t34 = *  HWINTRINSIC simd16 short Max REG NA <l:$2c4, c:$2c5>
                                                                        /--*  t73    mask   
                                                                        +--*  t34    simd16 
                                                                        +--*  t74    simd16 
Generating: N043 (???,???) [000075] -A-XG------                   t75 = *  HWINTRINSIC simd16 short ConditionalSelect REG d16
							V07 in reg d16 is becoming dead  [000055]
							Live regs: 0001000000000001 {x0 d16} - {d16} => 0000000000000001 {x0}
							Live vars after [000055]: {V03 V07} -{V07} => {V03}
```

So, destination of `ConditionalSelect+Max` is `d16`, and both inputs to `Max` are `d16`.


This can be fixed by removing the `prefUseNode` from the call to `BuildDelayFreeUses()` - see latest version of the PR. Now what happens is:

```
Generating: N029 (  4,  3) [000026] ---XG+-----                   t26 = *  IND       simd16 REG d16 <l:$2c1, c:$2c2>
IN000a:             ldr     q16, [x0, #0x08]
                                                                        /--*  t26    simd16 
Generating: N031 (  4,  3) [000052] DA-XG------                         *  STORE_LCL_VAR simd16 V07 cse0         d:1 d16 REG d16 $VN.Void
							V07 in reg d16 is becoming live  [000052]
							Live regs: 0000000000000001 {x0} + {d16} => 0001000000000001 {x0 d16}
							Live vars after [000052]: {V03} +{V07} => {V03 V07}
Generating: N033 (  1,  1) [000053] -----------                   t53 =    LCL_VAR   simd16 V07 cse0         u:1 d16 REG d16 <l:$2c1, c:$2c2>
Generating: N035 (  1,  1) [000055] -----------                   t55 =    LCL_VAR   simd16 V07 cse0         u:1 d16 (last use) REG d16 <l:$2c0, c:$300>
Generating: N037 (???,???) [000073] -----------                   t73 =    HWINTRINSIC mask   short CreateTrueMaskAll REG p0
IN000b:             ptrue   p0.h
Generating: N039 (???,???) [000074] -c---------                   t74 =    CNS_VEC   simd16<0x00000000, 0x00000000, 0x00000000, 0x00000000> REG NA
                                                                        /--*  t53    simd16 
                                                                        +--*  t55    simd16 
Generating: N041 (  7,  6) [000034] -A-XG+-----                   t34 = *  HWINTRINSIC simd16 short Max REG NA <l:$2c4, c:$2c5>
                                                                        /--*  t73    mask   
                                                                        +--*  t34    simd16 
                                                                        +--*  t74    simd16 
Generating: N043 (???,???) [000075] -A-XG------                   t75 = *  HWINTRINSIC simd16 short ConditionalSelect REG d17
							V07 in reg d16 is becoming dead  [000055]
							Live regs: 0001000000000001 {x0 d16} - {d16} => 0000000000000001 {x0}
							Live vars after [000055]: {V03 V07} -{V07} => {V03}
IN000c:             movprfx z17.h, p0/z, z16.h
IN000d:             smax    z17.h, p0/m, z17.h, z16.h
```

The same, but destination of `ConditionalSelect+Max` is now `d17`. The jit can now do the `movprfx` from `d16` to `d17` and also reuse `d16` in the `smax`.

The final assembly is correct. However what I'm unsure about is that if the lsra code is now correct, or if it's a fluke that it works.

(For reference, [failing jit dump](https://gist.github.com/a74nh/d094c16c57166195302ff7cdb53f2d1f) and [fixed jit dump](https://gist.github.com/a74nh/552e52f1e5692530a138232b276592c3))

jakobbotsch replied:
The ref positions created for both uses of `V07` are marked as last uses which is a bit surprising to me. It means that we will free the register at this location and also at the next location.
Even if they weren't, if we end up with the non-delay free ref position as the last use, it would end up freeing the register at this location. I don't know whether that is possible or not; I don't really follow the logic that builds these ref positions.
It's possible the main allocation logic here has a bug and that it needs to be paying special attention to `regsInUseNextLocation` as part of adding to the `regsToFree` set. However, both RPs getting `last` is definitely strange.

The safest fix for now would definitely be to consider all operands as delay freed.

jakobbotsch replied:
One thing I don't quite follow: why can we preference the target to be `op1` if we are going to emit `movprfx` that does not allow the destination register to be the same as `op1`?

jakobbotsch replied:
> The ref positions created for both uses of `V07` are marked as last uses which is a bit surprising to me.

Ah, that's a red herring, it's simply because when we print the first RP it is the last one, so at that point it is marked "last".

So indeed, passing `rmwNode` to `BuildDelayFreeUses` in this case makes it skip setting that ref position as delay-freed. I think that is an optimization; for example, it allows `x = x + x` to end up as `add rax, rax` on xarch. Without the optimization the destination would not be allowed to be `rax`.
But I suppose this optimization is just illegal in this case.

The only thing I'm unsure about is whether or not the operand we pick as the "delay" freed one is always going to be the last operand. Otherwise we would run into the problem I described above, where the ref position marked "last" is not the delay-freed one, causing us to free the register at the use's location.

tannergooding replied:
> One thing I don't quite follow: why can we preference the target to be op1 if we are going to emit movprfx that does not allow the destination register to be the same as op1?

The encoding requires `dst == op1`, because its RMW. You can't individually encode the separate registers. The `movprfx` then further requires that `op2 != dst`, there is no allowance for `dst == op1 == op2`.

tannergooding replied:
The simpler fix may be to handle `dst == op1 == op2` and emit the `ins; sel` pattern instead of `movprfx; sel`. Or to just never emit `mvprfx` in .NET 9

But if there is an easy way to guarantee that `dst != op2` when we know `movprfx` will be emitted (or we pessimize to assume it will for the .NET 9 fix), then that is the more ideal long term solution.

a74nh replied:
> But if there is an easy way to guarantee that `dst != op2`

Is that not covered by simply not passing `prefUseNode` to `BuildDelayFreeUses()` - as it's done in the latest version of the PR?
`srcCount += BuildDelayFreeUses(embOp2Node->Op(argNum));`

I do think there needs some better testing around this area, which is why I've added some extra hwintrinsic tests as part of #107084

jakobbotsch replied:
Yes, the delay free should accomplish it. The only concern I have is whether the delay free will always end up taking effect. I think you should audit the code and verify by inspection that, if the operands refer to the same interval (local), then the last ref position created will always be the one to be marked delay-free. A comment that there is a subtle requirement about that would be good too.

Also, it would be a good idea to inspect the other places that build delay free uses because of hard requirements to verify that they do not have this same bug around `rmwNode`.

a74nh replied:
Because we're not passing in a rmwnode to `BuildDelayFreeUses()` then the op2, op3 etc will always be marked as delay free.
I've added some assert checks to confirm this.
I've also merged it with the very similar code for FMA to remove duplication.

I'm not sure what would happen if the two locals were passed to the hwintrinsic in a different order, eg `55` then `52`
```
[000053] -----------                   t53 =    LCL_VAR   simd16 V07 cse0         u:1 d16 REG d16 <l:$2c1, c:$2c2>
[000055] -----------                   t55 =    LCL_VAR   simd16 V07 cse0         u:1 d16 (last use) REG d16 <l:$2c0, c:$300>
                                             /--*  t55    simd16 
                                             +--*  t53    simd16 
[000034] -A-XG+-----                   t34 = *  HWINTRINSIC simd16 short Max REG NA <l:$2c4, c:$2c5>
```
But I'm not sure that's possible.

jakobbotsch replied:
The order of them does not matter -- LSRA computes the precise last use information based on the ref positions, hence the last ref position created will be the one to be considered the "last use". For a last use LSRA will then free the register at location N + 1, or if the ref position is also marked as delay-freed, at location N + 2. Every node generally has its uses happen at location N and its definition happen at location N + 1.
So the only problem would be if we ended up creating the ref positions such that we ended up with two ref positions 
```
RP <not last> <delay free>
RP <last> <not delay free>
```

I think LSRA would free the register at location N + 1 in this case. To be honest that seems like a bug in the allocation algorithm, but not one I would be comfortable fixing for .NET 9 at this stage.

I think we are good with the fix to just not pass `rmwNode` now that you've went through some of the other cases and merged the code.

a74nh replied:
Opened #107134 to improve this code

kunalspathak replied:
I will sync-up with @a74nh offline about the general problem and right way to fix it. For now, the current fix sounds reasonable.

a74nh committed:
Fix comment

a74nh committed:
Add test header

a74nh committed:
don't delay prefUseOpNum

a74nh committed:
Fix FMA

a74nh committed:
Add assert checks for delay free

a74nh committed:
Merge embedded op build code

a74nh committed:
fix formatting

jakobbotsch commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -2045,22 +2032,38 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
+                srcCount += BuildDelayFreeUses(node, nullptr, RBM_NONE, &useRefPosition);
+
+#if defined(DEBUG)
+                // Ensure that if this node and the RMW node refer to the same local variable, then this
+                // node must be marked as delay free.
+                if (isCandidateLocalRef(node) && isCandidateLocalRef(prefUseNode) &&
+                    (getIntervalForLocalVarNode(node->AsLclVar()) ==
+                     getIntervalForLocalVarNode(prefUseNode->AsLclVar())))
                 {
-                    srcCount += BuildDelayFreeUses(embOp2Node->Op(argNum), prefUseNode);
+                    assert(useRefPosition->delayRegFree);
                 }
+#endif // defined(DEBUG)
```
Can it be just
```
// It is a hard requirement that these are not allocated to the same register as the destination,
// so verify no optimizations kicked in to skip setting the delay-free.
assert(useRefPosition->delayRegFree);
```
? The requirements apply equally to all operands, not just in the "same local" cases.

a74nh replied:
> ? The requirements apply equally to all operands, not just in the "same local" cases.

Done as suggested. Had to add an check that a use was actually created.

a74nh committed:
simplify assert

jakobbotsch commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -1984,21 +1985,7 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
-            }
-
-            srcCount += BuildDelayFreeUses(intrin.op3, emitOp1);
+            prefUseNode = emitOp1;
```
Don't we have the problem here for the `resultOpNum == 3` case? `prefUseNode` will end up being `intrinEmb.op3`, hence the last ref position we build can end up being a non-delay freed version of an earlier operand.

a74nh replied:
AIUI, the code in `hwintrinsiccodegenarm64.cpp` will choose the correct instruction based on which input register matches the output register.

For the `resultOpNum == 3` case it knows `op3` is the last use so ideally that should be the RMW operand. Then lsra ensures that is the one that does _not_ get delay freed.

Possible there is something I'm missing here.


If that logic is correct, then I think the code can be simplified to:

```c++
            unsigned resultOpNum =
                embOp2Node->GetResultOpNumForRmwIntrinsic(user, intrinEmb.op1, intrinEmb.op2, intrinEmb.op3);

            if (resultOpNum == 0)
            {
                   prefUseNode = embOp2Node->Op(1);
            }
            else
            {
                  assert(resultOpNum >= 1 && resultOpNum <= 3);
                 prefUseNode = embOp2Node->Op(resultOpNum);
            }
```

jakobbotsch replied:
What I'm not understanding is _when_ do we have the correctness requirement that the destination register is not the same as an operand register? Does this case have the correctness requirement?

I am less interesting in the optimizations happening here until we get the constraints encoded correctly in a way that I know register allocation will handle. Currently I am not convinced, but I also do not have a good understanding of what cases have the register requirements we have been discussing.

a74nh replied:
The requirements here would be that 1) `op1` and the destination are _ideally_ same register for RMW use 2) The other inputs can't use the same operand as the destination (because it might follow a `MOVPRFX`)

The codegen then outputs extra code to handle the case when `op1` and the destination don't match. Plus it asserts if other inputs share the same register as the destination.

The optimisation is that any of the inputs can be chosen as the RMW node (which is what the `GetResultOpNumForRmwIntrinsic()` is picking). Codegeneration spots this has happened by looking at which input matches the destination.

jakobbotsch replied:
Does anything prevent `prefUseNode` to be `op3` and `op2` to be the same interval as `op3`? I will create a delay-freed `op2` which has no effect. It will create non delay-freed `op3`. Then the destination register is allowed to be the same as `op2` and `op3`. If I understand correctly, this will hit the same problem as in this PR.

a74nh replied:
```
 var vr3 = vr2.F1;
 var vr4 = vr2.F1;
 var vr5 = vr2.F1;
 vr2.F1 = Sve.FusedMultiplyAdd(vr3, vr4, vr5);
```

```
Generating: N033 (  1,  1) [000060] -----------                   t60 =    LCL_VAR   simd16 V08 cse0         u:1 d16 REG d16 <l:$2c1, c:$2c2>
Generating: N035 (  1,  1) [000062] -----------                   t62 =    LCL_VAR   simd16 V08 cse0         u:1 d16 REG d16 <l:$2c0, c:$300>
Generating: N037 (  1,  1) [000063] -----------                   t63 =    LCL_VAR   simd16 V08 cse0         u:1 d16 (last use) REG d16 <l:$2c0, c:$300>
Generating: N039 (???,???) [000081] -----------                   t81 =    HWINTRINSIC mask   double CreateTrueMaskAll REG p0
IN000b:             ptrue   p0.d
Generating: N041 (???,???) [000082] -c---------                   t82 =    CNS_VEC   simd16<0x00000000, 0x00000000, 0x00000000, 0x00000000> REG NA
                                                                        /--*  t60    simd16
                                                                        +--*  t62    simd16
                                                                        +--*  t63    simd16
Generating: N043 (  8,  7) [000039] -A-XG+-----                   t39 = *  HWINTRINSIC simd16 double FusedMultiplyAdd REG NA <l:$2c5, c:$2c6>
                                                                        /--*  t81    mask
                                                                        +--*  t39    simd16
                                                                        +--*  t82    simd16
Generating: N045 (???,???) [000083] -A-XG------                   t83 = *  HWINTRINSIC simd16 double ConditionalSelect REG d17
							V08 in reg d16 is becoming dead  [000063]
							Live regs: 0001000000000001 {x0 d16} - {d16} => 0000000000000001 {x0}
							Live vars after [000063]: {V03 V08} -{V08} => {V03}
IN000c:             movprfx z17.d, p0/z, z16.d
IN000d:             fmla    z17.d, p0/m, z16.d, z16.d
```

This works because the codegeneration for `FusedMultiplyAdd` can handle cases where the registers are the same.

That wasn't the case for the non-FMA intrinsic in the original fuzzlyn test.

a74nh replied:
Which means we've got an inconsistency between different types of intrinsics in the codegen. Feels like the non-FMA intrinsic needs to be handling these cases as well? And the lsra should just be trying to give the most optimal register layout? ie the lsra doesn't have a correctness requirement?

a74nh replied:
Aha, that's because the conditionalselect node which wraps it has it's own destination register:
```
[000083] -A-XG------                   t83 = *  HWINTRINSIC simd16 double ConditionalSelect REG d17
``

And the `FusedMultiplyAdd` is contained within the `ConditionalSelect`, so codegen uses the target register of the  `ConditionalSelect`.

a74nh replied:
Curiously, this is exactly the two issues in #106867. Which this PR now fixes:

With the first:
```
Generating: N017 (  1,  1) [000028] -----------                   t28 =    LCL_VAR   simd16 V07 cse0         u:1 d17 REG d17 <l:$280, c:$2c0>
Generating: N019 (  3,  2) [000005] -----+-----                    t5 =    CNS_VEC   simd16<0x3f800000, 0x00000000, 0x00000000, 0x00000000> REG d18 $181
IN0009:             ldr     q18, (LARGELDC)[@RWD16]
Generating: N021 (  1,  1) [000030] -----------                   t30 =    LCL_VAR   simd16 V07 cse0         u:1 d17 (last use) REG d17 <l:$280, c:$2c0>
Generating: N023 (???,???) [000033] -----------                   t33 =    HWINTRINSIC mask   float CreateTrueMaskAll REG p1
IN000a:             ptrue   p1.s
Generating: N025 (???,???) [000034] -c---------                   t34 =    CNS_VEC   simd16<0x00000000, 0x00000000, 0x00000000, 0x00000000> REG NA
                                                                        /--*  t28    simd16 
                                                                        +--*  t5     simd16 
                                                                        +--*  t30    simd16 
Generating: N027 ( 12, 19) [000012] -A--G+-----                   t12 = *  HWINTRINSIC simd16 float FusedMultiplySubtractNegated REG NA <l:$300, c:$301>
                                                                        /--*  t33    mask   
                                                                        +--*  t12    simd16 
                                                                        +--*  t34    simd16 
Generating: N029 (???,???) [000035] -A--G------                   t35 = *  HWINTRINSIC simd16 float ConditionalSelect REG d17
							V07 in reg d17 is becoming dead  [000030]
							Live regs: 0002000000000001 {x0 d17} - {d17} => 0000000000000001 {x0}
							Live vars after [000030]: {V00 V07} -{V07} => {V00}
IN000b:             movprfx z17.s, p1/z, z17.s
IN000c:             fnmsb   z17.s, p1/m, z18.s, z17.s
```

is now:
```
Generating: N017 (  1,  1) [000028] -----------                   t28 =    LCL_VAR   simd16 V07 cse0         u:1 d17 REG d17 <l:$280, c:$2c0>
Generating: N019 (  3,  2) [000005] -----+-----                    t5 =    CNS_VEC   simd16<0x3f800000, 0x00000000, 0x00000000, 0x00000000> REG d18 $181
IN0009:             ldr     q18, (LARGELDC)[@RWD16]
Generating: N021 (  1,  1) [000030] -----------                   t30 =    LCL_VAR   simd16 V07 cse0         u:1 d17 (last use) REG d17 <l:$280, c:$2c0>
Generating: N023 (???,???) [000033] -----------                   t33 =    HWINTRINSIC mask   float CreateTrueMaskAll REG p1
IN000a:             ptrue   p1.s
Generating: N025 (???,???) [000034] -c---------                   t34 =    CNS_VEC   simd16<0x00000000, 0x00000000, 0x00000000, 0x00000000> REG NA
                                                                        /--*  t28    simd16 
                                                                        +--*  t5     simd16 
                                                                        +--*  t30    simd16 
Generating: N027 ( 12, 19) [000012] -A--G+-----                   t12 = *  HWINTRINSIC simd16 float FusedMultiplySubtractNegated REG NA <l:$300, c:$301>
                                                                        /--*  t33    mask   
                                                                        +--*  t12    simd16 
                                                                        +--*  t34    simd16 
Generating: N029 (???,???) [000035] -A--G------                   t35 = *  HWINTRINSIC simd16 float ConditionalSelect REG d19
							V07 in reg d17 is becoming dead  [000030]
							Live regs: 0002000000000001 {x0 d17} - {d17} => 0000000000000001 {x0}
							Live vars after [000030]: {V00 V07} -{V07} => {V00}
IN000b:             movprfx z19.s, p1/z, z17.s
IN000c:             fnmls   z19.s, p1/m, z18.s, z17.s
```

The second one: 
```
Generating: N017 (  1,  1) [000003] -----+-----                    t3 =    LCL_VAR   simd16<System.Numerics.Vector`1> V00 loc0         u:2 d16 REG d16 $100
Generating: N019 (  1,  1) [000004] -----+-----                    t4 =    LCL_VAR   simd16<System.Numerics.Vector`1> V00 loc0         u:2 d16 REG d16 $100
Generating: N021 (  1,  1) [000005] -----+-----                    t5 =    LCL_VAR   simd16<System.Numerics.Vector`1> V00 loc0         u:2 d16 (last use) REG d16 $100
Generating: N023 (???,???) [000025] -----------                   t25 =    HWINTRINSIC mask   short CreateTrueMaskAll REG p1
IN0006:             ptrue   p1.h
Generating: N025 (???,???) [000026] -c---------                   t26 =    CNS_VEC   simd16<0x00000000, 0x00000000, 0x00000000, 0x00000000> REG NA
                                                                        /--*  t3     simd16 
                                                                        +--*  t4     simd16 
                                                                        +--*  t5     simd16 
Generating: N027 (  4,  4) [000006] -c---+-----                    t6 = *  HWINTRINSIC simd16 short MultiplyAdd REG NA $200
                                                                        /--*  t25    mask   
                                                                        +--*  t6     simd16 
                                                                        +--*  t26    simd16 
Generating: N029 (???,???) [000027] -----------                   t27 = *  HWINTRINSIC simd16 short ConditionalSelect REG d16
							V00 in reg d16 is becoming dead  [000005]
							Live regs: 0001000000000000 {d16} - {d16} => 0000000000000000 {}
Debug: Closing V00 debug range.
							Live vars after [000005]: {V00} -{V00} => {}
IN0007:             movprfx z16.h, p1/z, z16.h
IN0008:             mad     z16.h, p1/m, z16.h, z16.h
```

is now:
```
Generating: N017 (  1,  1) [000003] -----+-----                    t3 =    LCL_VAR   simd16<System.Numerics.Vector`1> V00 loc0         u:2 d16 REG d16 $100
Generating: N019 (  1,  1) [000004] -----+-----                    t4 =    LCL_VAR   simd16<System.Numerics.Vector`1> V00 loc0         u:2 d16 REG d16 $100
Generating: N021 (  1,  1) [000005] -----+-----                    t5 =    LCL_VAR   simd16<System.Numerics.Vector`1> V00 loc0         u:2 d16 (last use) REG d16 $100
Generating: N023 (???,???) [000025] -----------                   t25 =    HWINTRINSIC mask   short CreateTrueMaskAll REG p1
IN0006:             ptrue   p1.h
Generating: N025 (???,???) [000026] -c---------                   t26 =    CNS_VEC   simd16<0x00000000, 0x00000000, 0x00000000, 0x00000000> REG NA
                                                                        /--*  t3     simd16 
                                                                        +--*  t4     simd16 
                                                                        +--*  t5     simd16 
Generating: N027 (  4,  4) [000006] -c---+-----                    t6 = *  HWINTRINSIC simd16 short MultiplyAdd REG NA $200
                                                                        /--*  t25    mask   
                                                                        +--*  t6     simd16 
                                                                        +--*  t26    simd16 
Generating: N029 (???,???) [000027] -----------                   t27 = *  HWINTRINSIC simd16 short ConditionalSelect REG d17
							V00 in reg d16 is becoming dead  [000005]
							Live regs: 0001000000000000 {d16} - {d16} => 0000000000000000 {}
Debug: Closing V00 debug range.
							Live vars after [000005]: {V00} -{V00} => {}
IN0007:             movprfx z17.h, p1/z, z16.h
IN0008:             mla     z17.h, p1/m, z16.h, z16.h
```

a74nh committed:
simplify FMA code

a74nh committed:
Add tests for 106867

a74nh committed:
ARM64-SVE: Allow op inside conditionselect to be non HWintrinsic

TEST_IMG: ubuntu/dotnet-build
TEST_CMD: safe ./projects/dotnet/build-runtime.sh

Jira: ENTLLT-7634
Change-Id: I337a291be6661f104fe90c7cdc27150eede43647

kunalspathak commented on src/tests/JIT/Regression/JitBlue/Runtime_106867/Runtime_106867.cs:
```diff
@@ -0,0 +1,51 @@
+{
+    public void M6(byte arg0, double arg1)
+    {
+        var vr0 = Vector128.CreateScalar(119.12962f).AsVector();
```
`if (Sve.IsSupported)`

kunalspathak commented on src/tests/JIT/Regression/JitBlue/Runtime_106867/Runtime_106867_1.cs:
```diff
@@ -0,0 +1,39 @@
+    [Fact]
+    public static void TestEntryPoint()
+    {
+        Vector<short> vr4 = default(Vector<short>);
```
`if (Sve.IsSupported)`

a74nh committed:
Add Sve.IsSupported to tests

a74nh committed:
Merge main

a74nh committed:
Add Sve.IsSupported to test

a74nh committed:
fix formatting

a74nh committed:
Revert "ARM64-SVE: Allow op inside conditionselect to be non HWintrinsic"

a74nh committed:
Revert "ARM64-SVE: Allow op inside conditionselect to be non HWintrinsic"

a74nh committed:
Revert "ARM64-SVE: Allow op inside conditionselect to be non HWintrinsic"

kunalspathak reviewed (approved):
LGTM. Thanks!

kunalspathak commented:
/backport to release/9.0-rc1

kunalspathak commented:
/backport to release/9.0
